### PR TITLE
feat: change agent to use v2 API for reporting stats

### PR DESCRIPTION
**Pull request by spikecurtis**
Modifies the agent to use the v2 API to report its statistics, using the `statsReporter` subcomponent.

### CODE REVIEW

File: agent/agenttest/client.go [hunk 1]
```diff
@@ -264,9 +228,13 @@ func (f *FakeAgentAPI) GetServiceBanner(context.Context, *agentproto.GetServiceB
...
-func (*FakeAgentAPI) UpdateStats(context.Context, *agentproto.UpdateStatsRequest) (*agentproto.UpdateStatsResponse, error) {
-	// TODO implement me
-	panic("implement me")
+func (f *FakeAgentAPI) UpdateStats(ctx context.Context, req *agentproto.UpdateStatsRequest) (*agentproto.UpdateStatsResponse, error) {
+	f.logger.Debug(ctx, "update stats called", slog.F("req", req))
+	// empty request is sent to get the interval; but our tests don't want empty stats requests
+	if req.Stats != nil {
+		f.statsCh <- req.Stats
+	}
+	return &agentproto.UpdateStatsResponse{ReportInterval: durationpb.New(statsInterval)}, nil
```
Comment (mtojek) [inline agent/agenttest/client.go:237]: Does it mean that tests are time dependent on real`statsInterval`? If so, it is possible to refactor them to use `clock`?

File: agent/agent_test.go [hunk 2]
```diff
@@ -177,14 +178,14 @@ func TestAgent_Stats_Magic(t *testing.T) {
 		require.Eventuallyf(t, func() bool {
 			s, ok := <-stats
 			t.Logf("got stats: ok=%t, ConnectionCount=%d, RxBytes=%d, TxBytes=%d, SessionCountVSCode=%d, ConnectionMedianLatencyMS=%f",
-				ok, s.ConnectionCount, s.RxBytes, s.TxBytes, s.SessionCountVSCode, s.ConnectionMedianLatencyMS)
+				ok, s.ConnectionCount, s.RxBytes, s.TxBytes, s.SessionCountVscode, s.ConnectionMedianLatencyMs)
```
Comment (mtojek) [inline agent/agent_test.go:181]: VSCode vs Vscode
PTY vs Pty

are these renamed on purpose?

File: agent/agent.go [hunk 3]
```diff
@@ -1218,115 +1223,83 @@ func (a *agent) handleReconnectingPTY(ctx context.Context, logger slog.Logger, m
...
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			duration, _, _, err := a.network.Ping(pingCtx, addresses[0].Addr())
+			if err != nil {
+				return
+			}
+			mu.Lock()
+			durations = append(durations, float64(duration.Microseconds()))
+			mu.Unlock()
```
Comment (mtojek) [inline agent/agent.go:1277]: could be deferred too

File: agent/agenttest/client.go (hunk 1 above)
Comment (spikecurtis) [reply agent/agenttest/client.go:237]: yes, they are, and yes, it's possible.  But, I think that's outside the scope of this PR.  I've just copied the hardcoded value from before into a `const`.

File: agent/agent_test.go (hunk 2 above)
Comment (spikecurtis) [reply agent/agent_test.go:181]: The SDK type fields were written by humans, the proto type fields are machine translated from lowercase_with_underscores so it doesn't get capitalization correct in all cases.  Unfortunate from a style perspective, but I haven't chased down how to fix it yet.

File: agent/agenttest/client.go (hunk 1 above)
Comment (mtojek) [reply agent/agenttest/client.go:237]: Oh, I assumed that it would be a low-hanging fruit to refactor, but I'm good with leaving for further.

**Review by mtojek — APPROVED** [review]
LGTM

**Comment by spikecurtis** [comment]
### Merge activity

* **Feb 7, 6:26 AM EST**: @spikecurtis merged this pull request with [Graphite](https://app.graphite.dev/github/pr/coder/coder/12024).

**Pull request merged**